### PR TITLE
Add statefulset partition controller

### PR DESCRIPTION
**Pull request by d-kuro**
refs: https://github.com/cybozu-go/moco/pull/628

See documentation for details:
https://github.com/cybozu-go/moco/blob/d-kuro/partition/docs/rolling-update-strategy.md

### CODE REVIEW

File: api/v1beta2/statefulset_webhhok.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+func SetupStatefulSetWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&appsv1.StatefulSet{}).
+		WithDefaulter(&StatefulSetDefaulter{}).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-apps-v1-statefulset,mutating=true,failurePolicy=fail,sideEffects=None,groups=apps,resources=statefulsets,verbs=create;update,versions=v1,name=statefulset.kb.io,admissionReviewVersions=v1
```
Comment (YZ775) [inline api/v1beta2/statefulset_webhhok.go:25]:
```suggestion
//+kubebuilder:webhook:path=/mutate-apps-v1-statefulset,mutating=true,failurePolicy=fail,sideEffects=None,groups=apps,resources=statefulsets,verbs=update,versions=v1,name=statefulset.kb.io,admissionReviewVersions=v1
```

I think this webhook does not use CREATE event., so I think required verbs is only `update`.

**Review by masa213f — CHANGES_REQUESTED** [review]
I could not update MySQL Pods after the following operations.
This means we can't roll back MySQLClusters by Argo CD, etc.

If the pod's phase is not `Running` (= `Pending`, `Succeeded`, or `Failed` phase), how about proceeding with the partition to update the error pod?
This behavior may be the same as the PodDisruptionBudget. Please see the following note.
https://kubernetes.io/docs/tasks/run-application/configure-pdb/#unhealthy-pod-eviction-policy
> NOTE: Pods in Pending, Succeeded or Failed phase are always considered for eviction.

1. Create a MySQLCluster
```
$ kubectl create ns test

$ kubectl apply -f - << EOF
apiVersion: moco.cybozu.com/v1beta2
kind: MySQLCluster
metadata:
  namespace: test
  name: test
spec:
  replicas: 3
  podTemplate:
    spec:
      containers:
      - name: mysqld
        image: ghcr.io/cybozu-go/moco/mysql:8.0.36.2
  volumeClaimTemplates:
  - metadata:
      name: mysql-data
    spec:
      accessModes: [ "ReadWriteOnce" ]
      resources:
        requests:
          storage: 1Gi
EOF
```

2. Wait to become Healthy.

3. Update the MySQLCluster with wrong image.

```
$ kubectl apply -f - << EOF
apiVersion: moco.cybozu.com/v1beta2
kind: MySQLCluster
metadata:
  namespace: test
  name: test
spec:
  replicas: 3
  podTemplate:
    spec:
      containers:
      - name: mysqld
        # invalid image
        image: foo/bar:baz
  volumeClaimTemplates:
  - metadata:
      name: mysql-data
    spec:
      accessModes: [ "ReadWriteOnce" ]
      resources:
        requests:
          storage: 1Gi
EOF
```

4. Wait for mysql pod `-2` to become error.

```
$ kubectl get pod -n test -w
NAME          READY   STATUS        RESTARTS   AGE
...
moco-test-2   3/3     Terminating   0          4m7s
...
moco-test-2   0/3     Pending       0          0s
...
moco-test-2   0/3     Init:ErrImagePull   0          17s
moco-test-2   0/3     Init:ImagePullBackOff   0          33s
```

File: api/v1beta2/statefulset_webhhok.go (hunk 1 above)
Comment (d-kuro) [reply api/v1beta2/statefulset_webhhok.go]: I fixed it:
https://github.com/cybozu-go/moco/blob/088bac5b85d159774d8888ad9baa992299794405/api/v1beta2/statefulset_webhhok.go#L25

**Comment by d-kuro** [comment]
@YZ775 @masa213f 
Sorry for the delay in fixing it.
I have significantly rewritten the code to check Pod ready when updating partition.
The `isRolloutReady` function has changed significantly.

https://github.com/cybozu-go/moco/blob/088bac5b85d159774d8888ad9baa992299794405/controllers/partition_controller.go#L134

And with that I have also added the case for E2E testing.
https://github.com/cybozu-go/moco/blob/088bac5b85d159774d8888ad9baa992299794405/e2e/partition_test.go#L163-L241

File: controllers/partition_controller.go [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+	}
+
+	patch := client.MergeFrom(sts.DeepCopy())
+	sts.Spec.UpdateStrategy.RollingUpdate.Partition = &newPartition
+
+	if err := r.Client.Patch(ctx, sts, patch); err != nil {
+		return fmt.Errorf("failed to patch new partition to StatefulSet %s/%s: %w", sts.Namespace, sts.Name, err)
+	}
+
+	r.Recorder.Eventf(sts, corev1.EventTypeNormal, "PartitionUpdate", "Updated partition from %d to %d", oldPartition, newPartition)
```
Comment (masa213f) [inline controllers/partition_controller.go:324]: Please output this information as log as well.

**Review by masa213f — CHANGES_REQUESTED** [review]
Please check and fix this behavior.
I am still reviewing, but I will comment first because this issue is important. :)

1. Create a MySQLCluster

```
$ kubectl create ns test

$ kubectl apply -f - << EOF
apiVersion: moco.cybozu.com/v1beta2
kind: MySQLCluster
metadata:
  namespace: test
  name: test
spec:
  replicas: 3
  podTemplate:
    spec:
      containers:
      - name: mysqld
        image: ghcr.io/cybozu-go/moco/mysql:8.0.36.2
  volumeClaimTemplates:
  - metadata:
      name: mysql-data
    spec:
      accessModes: [ "ReadWriteOnce" ]
      resources:
        requests:
          storage: 1Gi
EOF
```

2. Wait to become Healthy.

3. Rollout restart MySQL StatefulSet. (It's OK to update MySQLCluster)

```
$ kubectl rollout restart sts -n test moco-test
```

At this time, even though the rolling update has not been finished, the `partition` become to 0.

```
$ kubectl get sts -n test moco-test -o json | jq .spec.updateStrategy
{
  "rollingUpdate": {
    "partition": 0
  },
  "type": "RollingUpdate"
}

$ kubectl get pod -n test
NAME          READY   STATUS        RESTARTS   AGE
moco-test-0   3/3     Running       0          2m32s
moco-test-1   3/3     Running       0          2m32s
moco-test-2   3/3     Terminating   0          2m32s

$ kubectl events -n test
LAST SEEN               TYPE      REASON                    OBJECT                                         MESSAGE
...
28s                     Normal    Killing                   Pod/moco-test-2                                Stopping container slow-log
28s                     Normal    Killing                   Pod/moco-test-2                                Stopping container agent
28s                     Normal    Killing                   Pod/moco-test-2                                Stopping container mysqld
28s                     Normal    PartitionUpdate           StatefulSet/moco-test                          Updated partition from 3 to 2
28s                     Normal    PartitionUpdate           StatefulSet/moco-test                          Updated partition from 2 to 1
28s                     Normal    PartitionUpdate           StatefulSet/moco-test                          Updated partition from 1 to 0
6s (x10 over 28s)       Normal    SuccessfulDelete          StatefulSet/moco-test                          delete Pod moco-test-2 in StatefulSet moco-test successful
6s                      Normal    Scheduled                 Pod/moco-test-2                                Successfully assigned test/moco-test-2 to moco-worker2
6s (x9 over 7s)         Normal    RecreatingTerminatedPod   StatefulSet/moco-test                          StatefulSet test/moco-test is recreating terminated Pod
```

If the PARTITION is 0, when a mysql-0 pod is accidentally delete during a rolling update, it is re-created with a new revesion.
If the order of updating the MySQL Pods is swapped, the data may be corrupted during the MySQL upgrade.

https://github.com/cybozu-go/moco/blob/v0.23.2/docs/upgrading.md?plain=1#L58-L60

**Comment by masa213f** [comment]
@d-kuro 
IMO, it would be easier to implement by using StatefulSet status and MySQLCluster status.
For example, how about the following logic.

```
if cluster.generation != cluster.status.reconcileInfo.generation {
	// In this case, the reconciliation of MySQLClusterReconciler has not been completed.
	// Wait until completion.
	return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
}

if sts.metadata.generation != sts.status.observedGeneration {
	// In this case, the reconciliation of StatefulSet controller (kube-controller-manager) has not been completed.
	// Wait until completion.
	return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
}

if sts.spec.replicas == sts.status.updatedReplicas {
	// In this case, a rolling update has been completed.
	// Update the partition to the initial value (`sts.spec.replicas`) and finish the reconciliation.
	updatePartition(sts.spec.replicas) // return error if conflicts occur
	return reconcile.Result{}, nil
}

if sts.spec.updateStrategy.rollingUpdate.partition == 0 {
	// In this case, just wait for mysql-pod-0 to update.
	return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
}

nextIndex := sts.spec.updateStrategy.rollingUpdate.partition - 1
if podList.Items[nextIndex].Labels[appsv1.ControllerRevisionHashLabelKey] == sts.status.updateRevision {
	// In this case, nextIndex pod is already updated. Just decrement the partition.
	// The nextIndex pod will not be restarted, so we need not check the cluster status.
	updatePartition(sts.spec.replicas) // return error if conflicts occur
	return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
}

if cluster is healthy (check cluster.status) || nextIndex pod is not running (if possible, check the pod is ready or not) {
	// Restart the nextIndex pod.
	updatePartition(nextIndex) // return error if conflicts occur
	return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
}

// Wait until the cluster become healhty.
return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
```

File: api/v1beta2/statefulset_webhhok_test.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+		r := makeStatefulSet()
+		err := k8sClient.Create(ctx, r)
+		Expect(err).NotTo(HaveOccurred())
+
+		err = k8sClient.Update(ctx, r)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(r.Spec.UpdateStrategy.RollingUpdate.Partition).To(Equal(r.Spec.Replicas))
+	})
+
+	It("should not set partition when forcing updating StatefulSet", func() {
```
[the remark this answers is not in the document]
Comment (d-kuro) [reply api/v1beta2/statefulset_webhhok_test.go:103]: I added check:
https://github.com/cybozu-go/moco/pull/633/commits/69a62a0ffa68b99ee3b3d2fb9a02143f74fb9cb1

File: controllers/partition_controller.go (hunk 2 above)
Comment (d-kuro) [reply controllers/partition_controller.go:311]: Thanks! added:
https://github.com/cybozu-go/moco/pull/633/commits/abecf3569bb6ed4a5bc22428f4adf100a9d6b7b4#diff-7177516db0c5ea51dd9a5d501f166358b43c80e7e9db27bff587bc357d404bc0R310

**Comment by d-kuro** [comment]
@masa213f 
Thank you for the review and proposed revisions. I have revised the logic based on your suggestions. Please check it.
https://github.com/cybozu-go/moco/pull/633/commits/abecf3569bb6ed4a5bc22428f4adf100a9d6b7b4

File: api/v1beta2/statefulset_webhhok_test.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+func deleteStatefulSet() error {
+	r := &appsv1.StatefulSet{}
+	err := k8sClient.Get(ctx, client.ObjectKey{Namespace: "default", Name: "test"}, r)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
```
Comment (masa213f) [inline api/v1beta2/statefulset_webhhok_test.go:64]: A blank line here is unnecessary.

```suggestion
	r := &appsv1.StatefulSet{}
	err := k8sClient.Get(ctx, client.ObjectKey{Namespace: "default", Name: "test"}, r)
	if apierrors.IsNotFound(err) {
		return nil
	}
	if err != nil {
		return err
	}
```

File: controllers/partition_controller.go [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+	return podList, nil
+}
+
+func (r *StatefulSetPartitionReconciler) areAllChildPodsRolloutReady(ctx context.Context, sts *appsv1.StatefulSet, sortedPodList *corev1.PodList) bool {
+	log := crlog.FromContext(ctx)
+
+	nextRolloutTarget := r.nextRolloutTargetIndex(sts)
+
+	// Proceed with the rollout for the next Pod to be rolled out, even if it is not Ready.
+	// We expect that the Pod'excludeNextRolloutTagetPodList state will improve by being updated through the rollout.
```
Comment (masa213f) [inline controllers/partition_controller.go:214]: `excludeNextRolloutTagetPodList` is removed. Please rewrite this commnet.

File: controllers/partition_controller.go [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+			log.Info("Container is not ready", "pod", pod.Name, "namespace", pod.Namespace, "container", c.Name)
+			if !c.Ready {
+				return false
+			}
+		}
+		for _, c := range pod.Status.ContainerStatuses {
+			log.Info("Container is not ready", "pod", pod.Name, "namespace", pod.Namespace, "container", c.Name)
+			if !c.Ready {
+				return false
+			}
```
Comment (masa213f) [inline controllers/partition_controller.go:238]: Please change the field names in the log.
The `name` and `namespace` fields are duplicateed.

```
{"level":"info","ts":"2024-08-15T08:59:15Z","msg":"Pod is not ready","controller":"statefulset","controllerGroup":"apps","controllerKind":"StatefulSet","StatefulSet":{"name":"moco-test","namespace":"test"},"namespace":"test","name":"moco-test","reconcileID":"26c9b749-9edf-4904-9bf0-0cfb38156b4c","name":"moco-test-1","namespace":"test"}
{"level":"info","ts":"2024-08-15T08:59:25Z","msg":"Container is not ready","controller":"statefulset","controllerGroup":"apps","controllerKind":"StatefulSet","StatefulSet":{"name":"moco-test","namespace":"test"},"namespace":"test","name":"moco-test","reconcileID":"b8c0118e-83dd-488d-9569-d96724b418db","pod":"moco-test-1","namespace":"test","container":"copy-moco-init"}
```

Please also ckeck the other logs.

**Review by masa213f — CHANGES_REQUESTED** [review]

**Review by masa213f — APPROVED** [review]
LGTM. Thank you.

**Pull request merged**